File: src-tauri/src/db/models/build_session.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

// ============================================================================
// BuildPhase -- lifecycle state of a build session
// ============================================================================

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(export)]
pub enum BuildPhase {
    Initializing,
    Analyzing,
    AwaitingInput,
    Resolving,
    DraftReady,
    Completed,
    Failed,
    Cancelled,
    Testing,
    TestComplete,
    Promoted,
}

impl BuildPhase {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Initializing => "initializing",
            Self::Analyzing => "analyzing",
            Self::AwaitingInput => "awaiting_input",
            Self::Resolving => "resolving",
            Self::DraftReady => "draft_ready",
            Self::Completed => "completed",
            Self::Failed => "failed",
            Self::Cancelled => "cancelled",
            Self::Testing => "testing",
            Self::TestComplete => "test_complete",
            Self::Promoted => "promoted",
        }
    }

    pub fn is_terminal(&self) -> bool {
        matches!(
            self,
            Self::Completed | Self::Failed | Self::Cancelled | Self::Promoted
        )
    }

    /// Parse a phase string (as stored in SQLite) back into a `BuildPhase`.
    /// Returns `None` for unknown values instead of silently mapping to `Failed`.
    pub fn from_str_value(s: &str) -> Option<Self> {
        match s {
            "initializing" => Some(Self::Initializing),
            "analyzing" => Some(Self::Analyzing),
            "awaiting_input" => Some(Self::AwaitingInput),
            "resolving" => Some(Self::Resolving),
            "draft_ready" => Some(Self::DraftReady),
            "completed" => Some(Self::Completed),
            "failed" => Some(Self::Failed),
            "cancelled" => Some(Self::Cancelled),
            "testing" => Some(Self::Testing),
            "test_complete" => Some(Self::TestComplete),
            "promoted" => Some(Self::Promoted),
            _ => None,
        }
    }

    /// Validate a phase transition. Returns `Ok(())` if the transition is allowed.
    pub fn validate_transition(&self, next: BuildPhase) -> Result<(), String> {
        // Any phase can transition to Failed or Cancelled
        if matches!(next, Self::Failed | Self::Cancelled) {
            return Ok(());
        }

        let allowed = match self {
            Self::Initializing => matches!(next, Self::Analyzing),
            Self::Analyzing => matches!(
                next,
                Self::Resolving | Self::AwaitingInput | Self::DraftReady
            ),
            Self::AwaitingInput => matches!(next, Self::Resolving),
            Self::Resolving => matches!(next, Self::AwaitingInput | Self::DraftReady),
            Self::DraftReady => matches!(next, Self::Testing | Self::Resolving | Self::Promoted),
            Self::Testing => matches!(next, Self::TestComplete | Self::DraftReady | Self::Testing),
            Self::TestComplete => matches!(next, Self::Testing | Self::Promoted),
            Self::Completed | Self::Failed | Self::Cancelled | Self::Promoted => false,
        };

        if allowed {
            Ok(())
        } else {
            Err(format!(
                "Invalid phase transition: {} -> {}",
                self.as_str(),
                next.as_str()
            ))
        }
    }
}

// ============================================================================
// BuildEvent -- discriminated union streamed to frontend via Channel
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
#[ts(export)]
pub enum BuildEvent {
    CellUpdate {
        session_id: String,
        cell_key: String,
        data: serde_json::Value,
        status: String,
    },
    Question {
        session_id: String,
        cell_key: String,
        question: String,
        options: Option<Vec<String>>,
        /// Set only when the question was mirrored from a v3
        /// clarifying_question with scope=connector_category. Carries the
        /// machine-token category the vault picker should filter by.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        connector_category: Option<String>,
        /// C7 — true when the v3 question carried `accepts_reference: true`,
        /// meaning the answering UI should expose the file/URL attachment
        /// affordance and the answer payload may carry a `reference`.
        #[serde(default, skip_serializing_if = "std::ops::Not::not")]
        accepts_reference: bool,
        /// C7 — true when the v3 question carried
        /// `accepts_webhook_source: true`, meaning the answering UI should
        /// expose a smee.io URL input. Used when the LLM picks the
        /// `webhook` trigger type for a capability.
        #[serde(default, skip_serializing_if = "std::ops::Not::not")]
        accepts_webhook_source: bool,
    },
    Progress {
        session_id: String,
        dimension: Option<String>,
        message: String,
        percent: Option<f32>,
        activity: Option<String>,
    },
    Error {
        session_id: String,
        cell_key: Option<String>,
        message: String,
        retryable: bool,
    },
    SessionStatus {
        session_id: String,
        phase: String,
        resolved_count: usize,
        total_count: usize,
    },

    // ------------------------------------------------------------------
    // v3 capability-framework events (additive; coexist with legacy events)
    //
    // The CLI's three-phase protocol (behavior_core → capability_enumeration →
    // per-capability resolution) streams these event types. The run_session
    // parser ALSO mirrors each v3 event into a legacy CellUpdate so the
    // existing 3×3 matrix UI keeps rendering during the migration window
    // (§3.8 + §4.5 of C4-build-from-scratch-v3-handoff.md).
    // ------------------------------------------------------------------
    /// Phase A output: persona behavior core (mission + identity + voice +
    /// principles + constraints + decision_principles + verbosity_default).
    BehaviorCoreUpdate {
        session_id: String,
        data: serde_json::Value,
        status: String,
    },

    /// Phase B output: list of capability drafts (id + title + summary + goal).
    /// Emitted ONCE before per-capability resolution begins. Further drafts
    /// from "+ Add" flows are delivered as additional events with the same
    /// type; the frontend appends/dedupes by capability id.
    CapabilityEnumerationUpdate {
        session_id: String,
        data: serde_json::Value,
        status: String,
    },

    /// Phase C output: one field resolution on one capability.
    /// `field` is one of: suggested_trigger, connectors, notification_channels,
    /// review_policy, memory_policy, event_subscriptions, input_schema,
    /// sample_input, tool_hints, use_case_flow, error_handling.
    CapabilityResolutionUpdate {
        session_id: String,
        capability_id: String,
        field: String,
        value: serde_json::Value,
        status: String,
    },

    /// Persona-wide resolution: tools / connectors / notification_channels_default /
    /// operating_instructions / tool_guidance / error_handling / core_memories.
    PersonaResolutionUpdate {
        session_id: String,
        field: String,
        value: serde_json::Value,
        status: String,
    },

    /// v3 clarifying question — scoped to mission, a capability, or a specific
    /// field within a capability. Replaces legacy per-dimension `Question`
    /// when the CLI is emitting v3 events.
    ClarifyingQuestionV3 {
        session_id: String,
        /// One of: "mission" | "capability" | "field" | "connector_category"
        scope: String,
        capability_id: Option<String>,
        field: Option<String>,
        question: String,
        options: Option<Vec<String>>,
        /// For `scope: "connector_category"` — the machine token category the
        /// frontend vault picker should filter connectors by (e.g. "storage",
        /// "messaging", "image_generation"). None for other scopes.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        category: Option<String>,
        /// C7 — when true, the answering UI surfaces a file/URL attachment
        /// affordance. The answer command resolves the reference server-side
        /// and prepends a fenced block to the answer text before piping to
        /// the CLI. See `engine::build_session::reference`.
        #[serde(default, skip_serializing_if = "std::ops::Not::not")]
        accepts_reference: bool,
        /// C7 — when true, the answering UI surfaces a smee.io URL input.
        /// Emitted when the LLM picks the `webhook` trigger type (rule 24).
        #[serde(default, skip_serializing_if = "std::ops::Not::not")]
        accepts_webhook_source: bool,
    },
}

// ============================================================================
// BuildSession -- SQLite row representation
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BuildSession {
    pub id: String,
    pub persona_id: String,
    pub phase: BuildPhase,
    /// JSON-encoded map of resolved cells (stored as TEXT in SQLite).
    pub resolved_cells: String,
    /// JSON-encoded pending question (stored as TEXT in SQLite).
    pub pending_question: Option<String>,
    /// JSON-encoded agent intermediate representation.
    pub agent_ir: Option<String>,
    /// JSON-encoded adoption questionnaire answers (answers + question metadata + credential bindings).
    pub adoption_answers: Option<String>,
    pub intent: String,
    pub error_message: Option<String>,
    pub cli_pid: Option<u32>,
    /// Original workflow JSON for import mode (null for pure intent mode).
    pub workflow_json: Option<String>,
    /// Pre-parsed AgentIR from frontend parser for import mode.
    pub parser_result_json: Option<String>,
    /// Build mode — `Some("interactive")` (default ask-the-user gate flow) or
    /// `Some("one_shot")` (autonomous: LLM resolves every gate, retries up to
    /// 3× on test failure, auto-promotes on success). NULL on legacy rows is
    /// treated as `interactive` at read time.
    pub mode: Option<String>,
    /// Companion chat session that originated this build, when applicable.
    /// Used by the BuildWatcher job to post a result message back into the
    /// chat's episode log on terminal phase. NULL when the session was
    /// started from the regular UI (not via Companion).
    pub companion_session_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

impl BuildSession {
    /// Parse the JSON-encoded `agent_ir` string into a typed `AgentIr` struct.
    /// Returns `None` if the field is absent or unparseable.
    pub fn parse_agent_ir(&self) -> Option<super::agent_ir::AgentIr> {
        self.agent_ir
            .as_deref()
            .and_then(|s| serde_json::from_str(s).ok())
    }
}

// ============================================================================
// UserAnswer -- input from frontend to resume a paused session
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct UserAnswer {
    pub cell_key: String,
    pub answer: String,
    /// Optional reference attachment for clarifying questions whose IR shape
    /// carries `accepts_reference: true`. C7 (2026-04-27). At most one of
    /// `path` / `url` / `inline_content` is populated; the answer command
    /// resolves whichever is set, fences the contents, and prepends them to
    /// `answer` before piping to the CLI subprocess. See
    /// `engine::build_session::reference` for the contract.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reference: Option<UserAnswerReference>,
    /// Optional webhook source for clarifying questions whose IR shape
    /// carries `accepts_webhook_source: true`. C7 (2026-04-28). The answer
    /// command appends a fenced "WEBHOOK SOURCE" block to the answer text
    /// so the LLM places the URL on the relevant trigger config's
    /// `smee_channel_url` field. Promote-time then auto-creates the
    /// `smee_relays` row.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub webhook_source: Option<UserWebhookSource>,
}

/// Webhook source attachment payload — captured during the build wizard
/// when the LLM picks `trigger_type: "webhook"` for a capability and asks
/// the user for a smee.io URL via a clarifying_question carrying
/// `accepts_webhook_source: true`. C7 (2026-04-28).
///
/// The frontend submits this via `answer_build_question`; the answer
/// command appends a fenced summary of the URL to the answer text so the
/// LLM can place it on the trigger config's `smee_channel_url` field.
/// Promote-time then reads the trigger config and auto-creates a
/// `smee_relays` row pointing at this persona.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct UserWebhookSource {
    /// `https://smee.io/<channel>` URL the user pasted (or created at
    /// smee.io/new and copied back).
    pub channel_url: String,
    /// Optional comma-separated `event_type` allowlist forwarded to the
    /// `smee_relays.event_filter` column. Empty / None means no filter
    /// (relay forwards every smee event to the persona).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub event_filter: Option<String>,
}

/// Reference attachment payload — either a local file path, a URL, or
/// caller-supplied inline text.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct UserAnswerReference {
    /// Local filesystem path (resolved by the file-picker dialog on the
    /// frontend). Mutually exclusive with `url` / `inline_content`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    /// HTTPS URL — fetched server-side with SSRF protection.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    /// Pre-loaded text content (e.g. from a paste). Server still applies
    /// the size cap.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub inline_content: Option<String>,
    /// Optional human-friendly name shown to the LLM. Defaults to filename
    /// (file), URL (url), or "pasted reference" (inline).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
}

// ============================================================================
// PersistedBuildSession -- frontend-friendly hydration payload
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
pub struct PersistedBuildSession {
    pub id: String,
    pub persona_id: String,
    pub phase: BuildPhase,
    pub resolved_cells: serde_json::Value,
    pub pending_question: Option<serde_json::Value>,
    pub agent_ir: Option<serde_json::Value>,
    pub adoption_answers: Option<serde_json::Value>,
    pub intent: String,
    pub error_message: Option<String>,
    /// Build mode — `"interactive"` or `"one_shot"`. NULL legacy rows are
    /// surfaced as `Some("interactive")` so the frontend can switch on a
    /// string without dealing with three-state logic.
    pub mode: Option<String>,
    /// Companion chat session that originated this build, when applicable.
    pub companion_session_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

impl PersistedBuildSession {
    /// Convert a raw `BuildSession` (with JSON strings) into a frontend-friendly
    /// `PersistedBuildSession` (with parsed `serde_json::Value` fields).
    pub fn from_session(s: &BuildSession) -> Self {
        Self {
            id: s.id.clone(),
            persona_id: s.persona_id.clone(),
            phase: s.phase,
            resolved_cells: serde_json::from_str(&s.resolved_cells).unwrap_or_default(),
            pending_question: s
                .pending_question
                .as_deref()
                .and_then(|q| serde_json::from_str(q).ok()),
            agent_ir: s
                .agent_ir
                .as_deref()
                .and_then(|ir| serde_json::from_str(ir).ok()),
            adoption_answers: s
                .adoption_answers
                .as_deref()
                .and_then(|a| serde_json::from_str(a).ok()),
            intent: s.intent.clone(),
            error_message: s.error_message.clone(),
            mode: Some(s.mode.clone().unwrap_or_else(|| "interactive".to_string())),
            companion_session_id: s.companion_session_id.clone(),
            created_at: s.created_at.clone(),
            updated_at: s.updated_at.clone(),
        }
    }
}

// ============================================================================
// UpdateBuildSession -- partial update payload for repo::update
// ============================================================================

#[derive(Debug, Clone, Default)]
pub struct UpdateBuildSession {
    pub phase: Option<String>,
    pub resolved_cells: Option<String>,
    pub pending_question: Option<Option<String>>,
    pub agent_ir: Option<Option<String>>,
    pub adoption_answers: Option<Option<String>>,
    pub error_message: Option<Option<String>>,
    pub cli_pid: Option<Option<u32>>,
    pub mode: Option<Option<String>>,
    pub companion_session_id: Option<Option<String>>,
}

File: src/api/agents/buildSession.ts
/**
 * Typed Tauri invoke wrappers for build session commands.
 *
 * These map 1:1 to the Rust commands registered in lib.rs invoke_handler.
 * The Channel parameter enables streaming BuildEvent updates from the backend.
 */
import { invokeWithTimeout } from "@/lib/tauriInvoke";
import { Channel } from "@tauri-apps/api/core";
import type {
  BuildEvent,
  BuildReference,
  BuildWebhookSource,
  PersistedBuildSession,
  BuildSessionSummary,
  PromoteBuildResult,
  TestReport,
} from "@/lib/types/buildTypes";

/**
 * Start a new build session for a persona. Returns the session ID.
 * Build events are streamed via the provided Channel.
 *
 * `mode` selects the gate-resolution strategy:
 *   - `"interactive"` (default) — ask-the-user clarifying flow.
 *   - `"one_shot"` — autonomous: LLM resolves every gate, retries test
 *     failures up to 3×, auto-promotes on success. Read-only Glyph view
 *     while running; OS notification + bell entry on terminal phase.
 *
 * `companionSessionId` links the build back to the Companion chat that
 * originated it so the BuildWatcher job can post a result message into
 * that chat's episode log on terminal phase.
 */
export async function startBuildSession(
  channel: Channel<BuildEvent>,
  personaId: string,
  intent: string,
  workflowJson?: string | null,
  parserResultJson?: string | null,
  language?: string | null,
  mode?: 'interactive' | 'one_shot' | null,
  companionSessionId?: string | null,
  context?: string | null,
): Promise<string> {
  return invokeWithTimeout<string>("start_build_session", {
    channel,
    personaId,
    intent,
    workflowJson: workflowJson ?? null,
    parserResultJson: parserResultJson ?? null,
    language: language ?? null,
    mode: mode ?? null,
    companionSessionId: companionSessionId ?? null,
    context: context ?? null,
  });
}

/** Submit an answer to a pending build question.
 *
 * `reference` is optional — supply when the question carried
 * `acceptsReference: true` and the user attached a file / URL / inline text.
 * The backend resolves the reference server-side (SSRF-safe URL fetch,
 * size cap, content-type guard) and prepends a fenced block to the answer
 * before piping it to the CLI subprocess. See
 * `src-tauri/src/engine/build_session/reference.rs`.
 *
 * `webhookSource` is optional — supply when the question carried
 * `acceptsWebhookSource: true` and the user attached a smee.io URL. The
 * backend appends a fenced WEBHOOK SOURCE block to the answer text so the
 * LLM places the URL on the trigger config; promote then auto-creates the
 * smee_relays binding (see `commands/design/build_sessions.rs::auto_create_smee_relays`).
 */
export async function answerBuildQuestion(
  sessionId: string,
  cellKey: string,
  answer: string,
  reference?: BuildReference | null,
  webhookSource?: BuildWebhookSource | null,
): Promise<void> {
  return invokeWithTimeout("answer_build_question", {
    sessionId,
    cellKey,
    answer,
    reference: reference ?? null,
    webhookSource: webhookSource ?? null,
  });
}

/** Cancel an in-progress build session. */
export async function cancelBuildSession(sessionId: string): Promise<void> {
  return invokeWithTimeout("cancel_build_session", { sessionId });
}

/**
 * Update the per-capability disabled-dims map on a build session row.
 * Pass `null` to clear the column; pass a JSON string
 * `{ [use_case_id]: GlyphDimension[] }` to set it. Drives the
 * SigilEditModal toggle in adoption + build modes.
 */
export async function updateBuildSessionDisabledDims(
  sessionId: string,
  disabledDimsJson: string | null,
): Promise<void> {
  return invokeWithTimeout("update_build_session_disabled_dims", {
    sessionId,
    disabledDimsJson,
  });
}

/** Get the active (non-completed, non-cancelled) build session for a persona. */
export async function getActiveBuildSession(
  personaId: string,
): Promise<PersistedBuildSession | null> {
  return invokeWithTimeout<PersistedBuildSession | null>(
    "get_active_build_session",
    { personaId },
  );
}

/** List build session summaries, optionally filtered by persona. */
export async function listBuildSessions(
  personaId?: string,
): Promise<BuildSessionSummary[]> {
  return invokeWithTimeout<BuildSessionSummary[]>("list_build_sessions", {
    personaId: personaId ?? null,
  });
}

/**
 * Test a build draft by executing each tool against its real API.
 * Returns a per-tool test report with HTTP status codes, latency, and errors.
 */
export async function testBuildDraft(
  sessionId: string,
  personaId: string,
): Promise<TestReport> {
  return invokeWithTimeout<TestReport>("test_build_draft", {
    sessionId,
    personaId,
  }, undefined, 180_000);
}

/**
 * Promote a build draft to production: updates the persona with enriched
 * prompt data and atomically creates tools, triggers, and connectors.
 *
 * `excludedUseCaseIds` (A-grade Phase 5b): capability ids the user
 * excluded via the preview panel. Match against the LLM-emitted id
 * (e.g. `uc_morning_digest`), NOT the post-promote UUID-rekeyed id.
 * Optional; missing/empty means "promote everything in agent_ir".
 */
export async function promoteBuildDraft(
  sessionId: string,
  personaId: string,
  excludedUseCaseIds?: string[],
): Promise<PromoteBuildResult> {
  return invokeWithTimeout<PromoteBuildResult>("promote_build_draft", {
    sessionId,
    personaId,
    excludedUseCaseIds: excludedUseCaseIds ?? null,
  });
}

// ---------------------------------------------------------------------------
// Dry-run / simulation (C7) — preview a capability before promoting.
// Backend: src-tauri/src/commands/design/build_simulate.rs
// ---------------------------------------------------------------------------

/** Minimal shape of a PersonaExecution row returned by simulate_build_draft.
 * Snake-case fields mirror what `serde` emits (no rename_all on PersonaExecution).
 */
export interface SimulatedExecution {
  id: string;
  persona_id: string;
  status: string;
  is_simulation?: boolean;
  use_case_id?: string | null;
  created_at?: string;
  updated_at?: string;
  // Other fields (input_data, output_data, error_message, ...) are present
  // but the dry-run UI doesn't currently consume them — kept open via index.
  [key: string]: unknown;
}

/** Bundled artefacts response from get_simulation_artefacts. CamelCase per the
 * Rust struct's `#[serde(rename_all = "camelCase")]`. */
export interface SimulationArtefacts {
  executionId: string;
  /** Manual reviews emitted during the simulated execution. Snake-case fields
   * inside each item mirror PersonaManualReview as serialised today. */
  reviews: Array<{
    id: string;
    execution_id: string;
    persona_id: string;
    title: string;
    description: string | null;
    severity: string;
    status: "pending" | "approved" | "rejected" | "resolved";
    reviewer_notes: string | null;
    use_case_id: string | null;
    created_at: string;
    updated_at: string;
    [key: string]: unknown;
  }>;
  /** Memories the agent stored during the simulated execution. */
  memories: Array<{
    id: string;
    persona_id: string;
    title: string;
    content: string;
    category: string | null;
    importance: number | null;
    created_at: string;
    [key: string]: unknown;
  }>;
}

/**
 * Run a capability against the draft persona's `agent_ir` without promoting.
 * Snapshots a `design_context` onto the persona row, dispatches via
 * `execute_persona_inner` with `is_simulation=true`, returns the execution row.
 *
 * Throws AppError::Validation if the session is in a phase earlier than
 * `draft_ready`, or if `useCaseId` doesn't match any UC in the draft IR.
 */
export async function simulateBuildDraft(
  sessionId: string,
  useCaseId: string,
  inputOverride?: string | null,
): Promise<SimulatedExecution> {
  return invokeWithTimeout<SimulatedExecution>(
    "simulate_build_draft",
    {
      sessionId,
      useCaseId,
      inputOverride: inputOverride ?? null,
    },
    undefined,
    180_000,
  );
}

/**
 * Fetch artefacts (manual reviews, memories) a single simulation execution
 * produced. Used by the dry-run preview panel.
 */
export async function getSimulationArtefacts(
  executionId: string,
): Promise<SimulationArtefacts> {
  return invokeWithTimeout<SimulationArtefacts>("get_simulation_artefacts", {
    executionId,
  });
}
